### PR TITLE
feat: modularize the architecture and refactor contracts for root identity + sub tokens 

### PR DESCRIPTION
**Pull request by khushal1512**
### Description
This PR introduces architectural changes to the codebase and is modular now. 

- the file structure is modular and `modular/` folder houses `EndorsementModule` and `FlaggingModule.sol` In the coming weeks `RecoveryModule` etc will be written here and then inherited by IdentitySystem.sol in `src/`
- `EndorsementModule.sol`: Handles all peer-to-peer endorsement logic.
- `FlagModule.sol`: Handles flagging and trust-rating mechanics.

Additionally, 
Before this PR, users could only mint a single, transferable, non-revocable NFT. This PR replaces that logic with the following capabilities:

- Hierarchical Identity: Users can now mint a Root Identity and subsequently mint Sub-Tokens (education, age, certificate, social media etc)

- Social Primitives: Integrated the ability for identities to actively endorse or flag other users/identities directly through the modularized system.

- Revocability and Burning: Introduced the ability for users to burn their sub-tokens, and revoke other users subTokens (if had endorsed earlier)

Fixes #79

###  Screenshots/Recordings:
image

TODO: If applicable, add screenshots or recordings that demonstrate the interface **before** and **after** the changes.

###  Additional Notes:
I tested the contract using anvil and verified that it runs and behaves as intended

### AI Usage Disclosure:

We encourage contributors to use AI tools responsibly when creating Pull Requests. While AI can be a valuable aid, it is essential to ensure that your contributions meet the task requirements, build successfully, include relevant tests, and pass all linters. Submissions that do not meet these standards may be closed without warning to maintain the quality and integrity of the project. Please take the time to understand the changes you are proposing and their impact. AI slop is strongly discouraged and may lead to banning and blocking. Do not spam our repos with AI slop.

Check one of the checkboxes below:

- [ ] This PR does not contain AI-generated code at all.
- [ ] This PR contains AI-generated code. I have read the [AI Usage Policy](https://github.com/AOSSIE-Org/Info/blob/main/AI-UsagePolicy.md) and this PR complies with this policy. I have tested the code locally and I am responsible for it.

I have used the following AI models and tools: TODO

## Checklist

- [x] My PR addresses a single issue, fixes a single bug or makes a single improvement.
- [x] My code follows the project's code style and conventions
- [x] If applicable, I have made corresponding changes or additions to the documentation
- [] If applicable, I have made corresponding changes or additions to tests
- [x] My changes generate no new warnings or errors
- [x] I have joined the [Discord server](https://discord.gg/hjUhu33uAn) and I will share a link to this PR with the project maintainers there
- [x] I have read the [Contribution Guidelines](../CONTRIBUTING.md)
- [x] Once I submit my PR, CodeRabbit AI will automatically review it and I will address CodeRabbit's comments.
- [x] I have filled this PR template completely and carefully, and I understand that my PR may be closed without review otherwise.

## Summary by CodeRabbit

* **New Features**
  * Added **IdentitySystem** with non-transferable, username-based root identities and controlled transferable sub-tokens.
  * Supports sub-token creation, transfer (with endorsement wipe + history), burning, endorsements, and manual/automatic flagging.
  * Added lookup and view helpers (root/sub enumeration and username resolution).
* **Refactor**
  * Removed the old **IdentityToken** contract and replaced it with new shared types, events, error handling, schema storage, endorsement, and flag modules.
* **Chores**
  * Updated deployment/config scripts to use **IdentitySystem**; enabled Solidity IR compilation by default and adjusted coverage reporting/profile naming.
* **Tests**
  * Reworked test suite for end-to-end **IdentitySystem** coverage, including transfers, endorsements, flagging, and edge cases.

### CODE REVIEW

**Comment by khushal1512** [comment]
The CI checks are running on the previously hard coded file names and import statements, I'll fix them right away

File: src/IdentitySystem.sol [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+            transferCount: 0
+        });
+
+        rootToSubTokenIds[rootId].push(subTokenId);
+        walletSubTokens[msg.sender].push(subTokenId);
+        transferHistory[subTokenId].push(msg.sender);
+
+        emit Events.SubTokenCreated(subTokenId, rootId, tokenName, tokenType);
+        return subTokenId;
+    }
```
[the remark this answers is not in the document]
Comment (khushal1512) [reply src/IdentitySystem.sol:132]: good catch!

File: src/modules/FlagModule.sol [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    // Auto-Flagging (endorsement revocation threshold)
+
+    function _checkFlaggingThresholdInternal(uint256 subTokenId) internal {
+        uint256 endorsementCount = _getOriginalEndorsementCount(subTokenId);
+
+        if (endorsementCount > 0 && !_autoFlagged[subTokenId]) {
+            uint256 revokedCount = _countRevokedEndorsements(subTokenId);
+
+            // Auto-flags if >= 1/3rd of original endorsements are revoked
+            if (revokedCount * 3 >= endorsementCount) {
```
Comment (rohans02) [inline src/modules/FlagModule.sol:43]: Since making a root identity is free and `endorseSubToken` is open to any root, it looks like one throwaway identity could endorse a sub-token and then immediately revoke — with `originalEndorsementCount == 1`, that's `1*3 >= 1`, which trips the auto-flag on its own. So a single actor could flag someone else's token cheaply. Is that the intended sensitivity, or would a minimum-endorsement floor before auto-flagging make sense here? Could easily be missing context on my end.

File: src/IdentitySystem.sol [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+        subTokens[subTokenId].activeEndorsementCount--;
+    }
+
+    function _incrementRevokedCount(uint256 subTokenId) internal override {
+        subTokens[subTokenId].revokedCount++;
+    }
+
+    ///         Bumps generation so all prior endorsements become stale,
+    ///         then zeros the cached counters for the new epoch.
+    function _resetSubTokenEndorsementCounters(uint256 subTokenId) internal override {
```
Comment (rohans02) [inline src/IdentitySystem.sol]: The generation bump resets the endorsement counters, but `isFlagged` / `flagCount` / `_autoFlagged` / `_hasRootFlagged` aren't touched. So a flag seems to persist across a transfer to a new holder (even though their endorsements are wiped), a token auto-flagged once can't auto-flag again in a later epoch, and there's no un-flag path anywhere. Is the flag meant to be a permanent/sticky mark, or should it reset alongside the endorsements? Just want to confirm it's deliberate.

File: test/IdentityToken.t.sol [hunk 4]
```diff
@@ -2,685 +2,1261 @@
...
 
-    function test_SetAttribute_SocialLinks() public {
+    function test_RevertIf_CreateSubToken_NoRoot() public {
         vm.prank(alice);
-        uint256 tokenId = identityToken.mint();
+        vm.expectRevert("No root identity");
+        identitySystem.createSubToken("GitHub", "social", bytes(""), "", 0);
+    }
 
+    function test_RevertIf_CreateSubToken_RootDeactivated() public {
```
Comment (rohans02) [inline test/IdentityToken.t.sol]: This one looks like it passes without asserting anything — no `vm.expectRevert` or check inside. The guard it targets (`require(rootIdentities[rootId].isActive, ...)` in `createSubToken`) also seems unreachable since nothing sets `isActive = false` yet. Might be worth either wiring up a deactivate path + a real assertion here, or dropping both until that feature exists?

File: src/IdentitySystem.sol (hunk 3 above)
Comment (khushal1512) [reply src/IdentitySystem.sol]: Yes we do not want users to get rid of flags by transferring them to other wallets even if they reset the count of endorsements.

File: src/modules/FlagModule.sol (hunk 2 above)
Comment (khushal1512) [reply src/modules/FlagModule.sol:43]: fixed, now the subtoken needs to have 20 endorsers as the threshold. thanks for the catch!

**Comment by KanishkSogani** [comment]
hey @khushal1512, have a look on the changes pointed out by copilot above.

File: src/IdentitySystem.sol [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+        _resetEndorsements(subTokenId);
+
+        // Record transfer
+        transferHistory[subTokenId].push(sendingTo);
+        subTokens[subTokenId].transferCount++;
+
+        // Update wallet indices
+        _removeFromWalletList(msg.sender, subTokenId);
+        walletSubTokens[sendingTo].push(subTokenId);
+
```
[the remark this answers is not in the document]
Comment (khushal1512) [reply src/IdentitySystem.sol:150]: It is by design, lets say there is an Institution which mints bachelors degrees and then transfers them to the respective student wallets.
college->mints subtoken(degree) - college is the holder right now
when college transfers the token to student(rightful owner/new holder of the token) 
the people who will endorse the student will look at the parentRootId which is college and trust the subtoken.

If we changed the parentRootId to students rootId people would think the student minted a fake subtoken. 


case 2 : Alice minted a vaccine subtoken because the hospital does not provide nft certificates. lets say alice now wants to organize all health record tokens in a different wallet, alice transfers the vaccine token to other wallet and the parentRootId = alice will help others track the actual rootId of the token even when in different wallet for verifiability.

File: script/HelperConfig.s.sol [hunk 6]
```diff
@@ -1,23 +1,17 @@
...
-        if (block.chainid == 31337) activeNetworkConfig = getAnvilConfig();
-        else activeNetworkConfig = getSepoliaConfig();
+    function getConfig() public view returns (NetworkConfig memory) {
+        return getAnvilConfig();
     }
 
     function getAnvilConfig() public pure returns (NetworkConfig memory) {
+        // Anvil default account #0
         return NetworkConfig({ deployerKey: 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 });
     }
```
[the remark this answers is not in the document]
Comment (khushal1512) [reply script/HelperConfig.s.sol]: Addressed!

**Comment by khushal1512** [comment]
@KanishkSogani addressed one of the copilot comments and left comment on the (Update RootId Copilot comment)

following changes have been made:
- endorsements have a time validity
- endorsements do not reset upon transfer, instead emits an event for client
- use of custom errors across smart contracts
- maximum endorsement duration is 10 years and minimum is 30 days (removed)
- updated the tests according to above changes
